File: sub-agent-router-mcp-server/src/catalog.ts
import fs from 'fs';
import path from 'path';
import { AgentRegistry } from './registry.js';
import { AgentSpec, CommandSpec, SkillSpec } from './types.js';
import { loadMarketplace } from './marketplace.js';
import { normalizeId } from './utils.js';

export interface CatalogOptions {
  registry: AgentRegistry;
  marketplacePath?: string;
  pluginsRoot?: string;
}

export class SubAgentCatalog {
  private registry: AgentRegistry;
  private marketplacePath?: string;
  private pluginsRoot?: string;
  private agents = new Map<string, AgentSpec>();
  private skills = new Map<string, SkillSpec>();
  private contentCache = new Map<string, string>();

  constructor(options: CatalogOptions) {
    this.registry = options.registry;
    this.marketplacePath = options.marketplacePath;
    this.pluginsRoot = options.pluginsRoot;
    this.reload();
  }

  reload() {
    this.agents.clear();
    this.skills.clear();
    const marketplace =
      this.marketplacePath && fs.existsSync(this.marketplacePath)
        ? loadMarketplace({ marketplacePath: this.marketplacePath, pluginsRoot: this.pluginsRoot })
        : { agents: [], skills: [] };

    for (const skill of marketplace.skills) {
      if (!this.skills.has(skill.id)) {
        this.skills.set(skill.id, skill);
      }
    }
    for (const agent of marketplace.agents) {
      this.agents.set(agent.id, agent);
    }

    const registryAgents = this.registry.listAgents();
    for (const agent of registryAgents) {
      this.agents.set(agent.id, agent);
    }
  }

  listAgents(): AgentSpec[] {
    return Array.from(this.agents.values());
  }

  getAgent(id: string): AgentSpec | null {
    const normalized = normalizeId(id);
    if (!normalized) return null;
    return this.agents.get(normalized) || null;
  }

  listSkills(): SkillSpec[] {
    return Array.from(this.skills.values());
  }

  getSkill(id: string): SkillSpec | null {
    const normalized = normalizeId(id);
    if (!normalized) return null;
    return this.skills.get(normalized) || null;
  }

  resolveSkills(skillIds: string[]): SkillSpec[] {
    const result: SkillSpec[] = [];
    for (const skillId of skillIds || []) {
      const normalized = normalizeId(skillId);
      if (!normalized) continue;
      const skill = this.skills.get(normalized);
      if (skill) result.push(skill);
    }
    return result;
  }

  resolveCommand(agent: AgentSpec, commandId?: string): CommandSpec | null {
    const commands = Array.isArray(agent.commands) ? agent.commands : [];
    if (!commandId) {
      return pickFirstCommand(commands, agent.defaultCommand);
    }
    const target = normalizeId(commandId).toLowerCase();
    return (
      commands.find((cmd) => normalizeId(cmd.id).toLowerCase() === target) ||
      commands.find((cmd) => normalizeId(cmd.name).toLowerCase() === target) ||
      null
    );
  }

  readContent(filePath?: string): string {
    if (!filePath) return '';
    const resolved = path.resolve(filePath);
    if (this.contentCache.has(resolved)) {
      return this.contentCache.get(resolved) || '';
    }
    let text = '';
    try {
      text = fs.readFileSync(resolved, 'utf8');
    } catch {
      text = '';
    }
    this.contentCache.set(resolved, text);
    return text;
  }

  setPluginsRoot(root?: string) {
    this.pluginsRoot = root;
    this.reload();
  }
}

function pickFirstCommand(commands: CommandSpec[], preferredId?: string): CommandSpec | null {
  if (!commands || commands.length === 0) return null;
  if (preferredId) {
    const target = normalizeId(preferredId).toLowerCase();
    const match =
      commands.find((cmd) => normalizeId(cmd.id).toLowerCase() === target) ||
      commands.find((cmd) => normalizeId(cmd.name).toLowerCase() === target);
    if (match) return match;
  }
  return commands[0];
}

File: sub-agent-router-mcp-server-rs/src/catalog.rs
use crate::marketplace::load_marketplace;
use crate::registry::AgentRegistry;
use crate::types::{AgentSpec, CommandSpec, SkillSpec};
use crate::utils::normalize_id;
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

pub struct SubAgentCatalog {
    registry: AgentRegistry,
    marketplace_path: Option<PathBuf>,
    plugins_root: Option<PathBuf>,
    agents: HashMap<String, AgentSpec>,
    skills: HashMap<String, SkillSpec>,
    content_cache: HashMap<String, String>,
}

impl SubAgentCatalog {
    pub fn new(registry: AgentRegistry, marketplace_path: Option<PathBuf>, plugins_root: Option<PathBuf>) -> Self {
        let mut catalog = Self {
            registry,
            marketplace_path,
            plugins_root,
            agents: HashMap::new(),
            skills: HashMap::new(),
            content_cache: HashMap::new(),
        };
        catalog.reload();
        catalog
    }

    pub fn reload(&mut self) {
        self.agents.clear();
        self.skills.clear();
        if let Some(path) = self.marketplace_path.clone() {
            let result = load_marketplace(&path, self.plugins_root.as_deref());
            for skill in result.skills {
                self.skills.entry(skill.id.clone()).or_insert(skill);
            }
            for agent in result.agents {
                self.agents.insert(agent.id.clone(), agent);
            }
        }
        for agent in self.registry.list_agents() {
            self.agents.insert(agent.id.clone(), agent);
        }
    }

    pub fn list_agents(&self) -> Vec<AgentSpec> {
        self.agents.values().cloned().collect()
    }

    pub fn get_agent(&self, id: &str) -> Option<AgentSpec> {
        let normalized = normalize_id(Some(&id.to_string()));
        if normalized.is_empty() {
            None
        } else {
            self.agents.get(&normalized).cloned()
        }
    }

    pub fn resolve_skills(&self, ids: &[String]) -> Vec<SkillSpec> {
        let mut result = Vec::new();
        for id in ids {
            let normalized = normalize_id(Some(id));
            if normalized.is_empty() {
                continue;
            }
            if let Some(skill) = self.skills.get(&normalized) {
                result.push(skill.clone());
            }
        }
        result
    }

    pub fn resolve_command(&self, agent: &AgentSpec, command_id: Option<&str>) -> Option<CommandSpec> {
        let commands = agent.commands.clone().unwrap_or_default();
        if commands.is_empty() {
            return None;
        }
        if let Some(target) = command_id {
            let normalized = normalize_id(Some(&target.to_string())).to_lowercase();
            if let Some(cmd) = commands
                .iter()
                .find(|c| normalize_id(Some(&c.id)).to_lowercase() == normalized)
            {
                return Some(cmd.clone());
            }
            if let Some(cmd) = commands.iter().find(|c| {
                c.name
                    .as_ref()
                    .map(|n| normalize_id(Some(n)).to_lowercase() == normalized)
                    .unwrap_or(false)
            }) {
                return Some(cmd.clone());
            }
        }
        pick_first_command(&commands, agent.default_command.as_deref())
    }

    pub fn read_content(&mut self, path: Option<&str>) -> String {
        let file_path = match path {
            Some(p) => p,
            None => return String::new(),
        };
        let resolved = Path::new(file_path).to_path_buf();
        let key = resolved.to_string_lossy().to_string();
        if let Some(cached) = self.content_cache.get(&key) {
            return cached.clone();
        }
        let text = fs::read_to_string(&resolved).unwrap_or_default();
        self.content_cache.insert(key, text.clone());
        text
    }

    pub fn set_plugins_root(&mut self, root: Option<PathBuf>) {
        self.plugins_root = root;
        self.reload();
    }
}

fn pick_first_command(commands: &[CommandSpec], preferred_id: Option<&str>) -> Option<CommandSpec> {
    if commands.is_empty() {
        return None;
    }
    if let Some(preferred) = preferred_id {
        let normalized = normalize_id(Some(&preferred.to_string())).to_lowercase();
        if let Some(cmd) = commands
            .iter()
            .find(|c| normalize_id(Some(&c.id)).to_lowercase() == normalized)
        {
            return Some(cmd.clone());
        }
        if let Some(cmd) = commands.iter().find(|c| {
            c.name
                .as_ref()
                .map(|n| normalize_id(Some(n)).to_lowercase() == normalized)
                .unwrap_or(false)
        }) {
            return Some(cmd.clone());
        }
    }
    commands.first().cloned()
}
